### PR TITLE
[COST-4741] - Separate AWS node network costs into a Network unattributed namespace

### PR DESCRIPTION
**Pull request by samdoran**
## Jira Ticket

[COST-4741](https://issues.redhat.com/browse/COST-4741)

## Description

Separates out AWS node network costs into a separate project for later distribution.

## Testing

1. Load OCP on AWS test data.
    This is currently broken. You will need to run the second command twice.

    ```
    make create-test-customer
    make load-test-customer-data test_source=aws
    ```
    
1. Inspect data in Trino. Where product_family is `Data Transfer` and `product_code` is `AmazonEC2`, there should be a data transfer direction

    ```
    SELECT namespace,product_code,product_family,data_transfer_direction,usage_amount FROM reporting_ocpawscostlineitem_project_daily_summary;
    ```
    
    ```
    namespace                 |  product_code  |   product_family   | data_transfer_direction |     usage_amount
    --------------------------+----------------+--------------------+-------------------------+-----------------------
     cost-management          | AmazonEC2      | Compute Instance   | NULL                    |                   8.0
     Network unattributed     | AmazonEC2      | Data Transfer      | IN                      |   0.05788159130273985
     Network unattributed     | AmazonEC2      | Data Transfer      | IN                      |   0.08385133566183305
     Network unattributed     | AmazonEC2      | Data Transfer      | OUT                     | 4.0166646410467517E-4
     Network unattributed     | AmazonEC2      | Data Transfer      | IN                      |  0.013915647933780239
     Platform unallocated     | AmazonEC2      | Data Transfer      | NULL                    |   0.04199137044014825
     Network unattributed     | AmazonEC2      | Data Transfer      | OUT                     |  0.050977435375557656
    ...
    ```

1. Verify there are records for Network unattributed in PostgreSQL AWS daily summary table

    ```
    SELECT * FROM reporting_ocpawscostlineitem_project_daily_summary_p WHERE namespace = 'Network unattributed';
    ```
    
1. Verify OpenShift costs are correct

    ```
    curl http://localhost:8000/api/cost-management/v1/reports/openshift/costs/
    ```
    
1. Verify AWS filtered by OpenShift costs are correct

    ```
    http://localhost:8000/api/cost-management/v1/reports/openshift/infrastructures/aws/costs/
    ```
    
1. Compare Trino and PostgreSQL data to make sure costs add up:

    ```
    postgres=# SELECT sum(unblended_cost) from reporting_aws_cost_summary_p
    sum
    -----------------
     23785.362736418
    
    trino:org1234567> SELECT sum(lineitem_unblendedcost) from aws_line_items;
    _col0
    --------------------
     23785.362736451105
    (1 row)
    ```
    

## Release Notes
- [x] proposed release note

```markdown
* [COST-3761](https://issues.redhat.com/browse/COST-3761) Separate AWS EC2 instance network costs into a `Network unattributed` project
  Network costs were previously combined with the AWS EC2 instance costs. Now those costs presented in a new `Network unattributed` project.
```

### CODE REVIEW

File: koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql [hunk 1]
```diff
@@ -198,6 +201,19 @@ SELECT cast(uuid() as varchar) as uuid,
...
+    -- Determine network direction
+    CASE
+        -- Is this a network record?
+        WHEN max(aws.lineitem_productcode) = 'AmazonEC2' AND max(aws.product_productfamily) = 'Data Transfer' THEN
+            -- Yes, it's a network record. What's the direction?
+            CASE
+                WHEN position('in-bytes' IN lower(max(aws.lineitem_usagetype))) > 0 THEN 'IN'
+                WHEN position('out-bytes' IN lower(max(aws.lineitem_usagetype))) > 0 THEN 'OUT'
+                WHEN (position('regional-bytes' IN lower(max(aws.lineitem_usagetype))) > 0 AND position('-in' IN lower(max(lineitem_operation))) > 0) THEN 'IN'
+                WHEN (position('regional-bytes' IN lower(max(aws.lineitem_usagetype))) > 0 AND position('-out' IN lower(max(lineitem_operation))) > 0) THEN 'OUT'
```
Comment (cgoodfred) [inline koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql:213]: any reason to use `position` over `strpos` when we've been using `strpos` everywhere else in TrinoSQL files?

File: koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql [hunk 2]
```diff
@@ -198,6 +201,19 @@ SELECT cast(uuid() as varchar) as uuid,
...
+                ELSE NULL
+            END
+    END AS data_transfer_direction,
```
Comment (cgoodfred) [inline koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql:216]: This SQL is grouped by these items below, are these enough to ensure a unique `data_transfer_direction` or should we be grouping by the `data_transfer_direction` result as well?
```
GROUP BY aws.lineitem_usagestartdate,
    aws.lineitem_resourceid,
    4, -- CASE satement
    aws.product_productfamily,
    aws.product_instancetype,
    aws.lineitem_availabilityzone,
    aws.product_region,
    aws.resourcetags,
    aws.costcategory
```

File: koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql [hunk 3]
```diff
@@ -717,6 +738,116 @@ LEFT JOIN postgres.{{schema | sqlsafe}}.reporting_awsaccountalias AS aa
...
+    month,
+    day
+)
+SELECT
+    aws.uuid AS aws_uuid,
+    max(cluster_id),
+    max(cluster_alias),
+    max(data_source),
+    'Network unattributed' AS namespace,
+    max(node),
```
Comment (cgoodfred) [inline koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql:792]: Don't need the max here, this is grouped by node.
```suggestion
    node,
```

File: koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql [hunk 4]
```diff
@@ -781,6 +915,15 @@ SELECT uuid(),
...
     usage_amount,
+    CASE
+        WHEN upper(data_transfer_direction) = 'IN' THEN usage_amount
+        ELSE 0
+    END AS infrastructure_data_in_gigabytes,
+    CASE
+        WHEN upper(data_transfer_direction) = 'OUT' THEN usage_amount
+        ELSE 0
+    END AS infrastructure_data_out_gigabytes,
+    data_transfer_direction,
```
Comment (cgoodfred) [inline koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql:926]: Just checking cause I'm not sure, are the units for the AWS data always in `GB`?

**Review by cgoodfred — COMMENTED** [review]
Just a few quick questions/comments

File: koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql (hunk 1 above)
Comment (samdoran) [reply koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql:213]: No reason other than it's easier to read.

Comment (maskarb) [reply koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql:213]: literally everywhere else we use strpos...

Comment (samdoran) [reply koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql:213]: Ok, I will switch to that function.

File: koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql (hunk 4 above)
Comment (samdoran) [reply koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql:926]: I looked at some real data and `pricing_unit` is always `GB` for data transfer related fields.

File: koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql (hunk 2 above)
Comment (samdoran) [reply koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql:216]: We do probably need to group by the direction, but we can't use an aggregate in a group by. Grouping by `aws.lineitem_usagetype` may work.

Comment (cgoodfred) [reply koku/masu/database/trino_sql/reporting_ocpawscostlineitem_daily_summary.sql:216]: You can also use numbers to group by items in the `SELECT` statement, for example the `4` in the group by there is based on the `CASE` statement in the SQL

**Review by lcouzens — DISMISSED** [review]

**Review by cgoodfred — APPROVED** [review]

**Review by cgoodfred — APPROVED** [review]

**Pull request merged**